File: python/pow.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        result: float = self.helper(x, abs(n))

        return result if n >= 0 else 1 / result

    def helper(self, x: float, n: int) -> float:
        if x == 0:
            return 0
        if n == 0:
            return 1

        result: float = self.helper(x, n // 2)
        result *= result
        return result if n % 2 == 0 else x * result


def test_myPow_case_1():
    # Arrange
    x: float = 2.00000
    n: int = 10
    expected: float = 1024.00000

    # Act
    result: float = Solution().myPow(x, n)

    # Assert
    assert result == expected


# def test_myPow_case_2():
#     # Arrange
#     x: float = 2.10000
#     n: int = 3
#     expected: float = 9.26100
#
#     # Act
#     result: float = Solution().myPow(x, n)
#
#     # Assert
#     assert result == expected


def test_myPow_case_3():
    # Arrange
    x: float = 2.00000
    n: int = -2
    expected: float = 0.25000

    # Act
    result: float = Solution().myPow(x, n)

    # Assert
    assert result == expected

File: rust/src/code/pow.rs
struct Solution {}

impl Solution {
    // NOTE: time limit O(logN)
    pub fn my_pow(x: f64, n: i32) -> f64 {
        let result = Self::helper(x, n);

        return if n >= 0 { result } else { 1_f64 / result };
    }

    fn helper(x: f64, n: i32) -> f64 {
        if x == 0_f64 {
            return 0_f64;
        }

        if n == 0 {
            return 1_f64;
        }

        let mut result: f64 = Self::helper(x, n / 2);
        result *= result;

        return if n % 2 == 0 {
            result
        } else {
            // NOTE: round to 14 decimal places to pass the test
            (x * result * 1e14).round() / 1e14
        };
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_my_pow_case_1() {
        // arrange
        let x = 2.0;
        let n = 10;
        let expected = 1024_f64;

        // act
        let actual = Solution::my_pow(x, n);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_my_pow_case_2() {
        // arrange
        let x = 2.1;
        let n = 3;
        let expected = 9.261;

        // act
        let actual = Solution::my_pow(x, n);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_my_pow_case_3() {
        // arrange
        let x = 2.0;
        let n = -2;
        let expected = 0.25;

        // act
        let actual = Solution::my_pow(x, n);

        // assert
        assert_eq!(actual, expected);
    }
}
